src/sql/azdata.proposed.d.ts

```diff
@@ -976,4 +976,168 @@ declare module 'azdata' {
+
+		export interface TableDesignerProvider extends DataProvider {
+			getTableDesignerInfo(table: TableInfo): Promise<TableDesignerInfo>;
+			processTableEdit(table: TableInfo, data: DesignerData, edit: DesignerEdit): Promise<DesignerEditResult>;
+		}
+
+		/**
+		 * The information of the table.
+		 */
+		export interface TableInfo {
+			server: string;
+			database: string;
+			schema?: string;
+			name?: string;
+			isNewTable: boolean;
+			[key: string]: any;
+		}
+
+		export interface TableDesignerInfo {
+			view: TableDesignerView;
+			data: DesignerData;
+			columnTypes: string[];
+			schemas: string[];
+		}
+
+		export enum TableProperty {
+			Columns = 'columns',
+			Description = 'description',
+			Name = 'name',
+			Schema = 'schema',
+			Script = 'script'
+		}
+
+		export enum TableColumnProperty {
+			AllowNulls = 'allowNulls',
+			DefaultValue = 'defaultValue',
+			Length = 'length',
+			Name = 'name',
+			Type = 'type'
+		}
+
+		export interface TableDesignerView {
+			additionalTableProperties?: DesignerComponentType[];
+			addtionalTableColumnProperties?: DesignerComponentType[];
+			addtionalTabs?: DesignerTab[];
+		}
+
+		export interface DesignerData {
+			[key: string]: InputComponentData | CheckboxComponentData | DropdownComponentData | TableComponentData;
+		}
+
+		export interface ComponentData {
+			enabled?: boolean;
+		}
+
+		export interface InputComponentData extends ComponentData {
+			value: string | number;
+		}
+
+		export interface CheckboxComponentData extends ComponentData {
+			value: boolean;
+		}
+
+		export interface DropdownComponentData extends ComponentData {
+			value: string;
+			options: string[];
+		}
+
+		export interface TableComponentData extends ComponentData {
+			rows: TableComponentRowData[];
+		}
+
+		export interface TableComponentRowData {
+			[key: string]: InputComponentData | CheckboxComponentData | DropdownComponentData | TableComponentData;
+		}
+
+		export enum DesignerEditType {
+			Add = 0,
+			Remove = 1,
+			Update = 2
+		}
+
+		export interface DesignerEdit {
+			type: DesignerEditType;
+			property: DesignerEditIdentifier;
+			value: any;
+		}
+
+		export type DesignerEditIdentifier = string | { parentProperty: string, index: number, property: string };
+
+		export interface DesignerEditResult {
+			data: DesignerData;
+			isValid: boolean;
+			errorMessages?: string[];
+		}
+
+		export interface DesignerTab {
+			title: string;
+			/**
+			 * Whether the label should be place on top of the component or on the left side of component. Default value is false.
+			 */
+			labelOnTop?: boolean;
+			components: DesignerComponentType[];
+		}
+
+		export type DesignerComponentType = InputComponentDefinition | CheckboxComponentDefinition | DropdownComponentDefinition | TableComponentDefinition;
+
+		export type DesignerComponentTypeName = 'input' | 'checkbox' | 'dropdown' | 'table';
+
+		export interface ComponentDefinition {
```

Comment:
I wonder if we can use the modelview component types directly in the designer UI descriptions?  It seems redundant to duplicate a lot of that logic into this namespace.

Reply:
+1, this is going to be annoying to maintain parity and if it's using modelview underneath anyways not sure why we'd want to hide options.

Reply:
But also - how important is it for us to handle custom UI like this? Couldn't we do it more like the resource deployment where it does all the UI stuff and contributors just give the stuff to fill in?